### PR TITLE
Instrument ReferenceTracker cleanup to try to catch double delete crashes

### PR DESCRIPTION
**Pull request by jevansaks**
We're seeing sporadic double delete crashes in the vicinity of ReferenceTracker's DeleteInstanceOnUIThread cleanup lambda. My best guess (supported by some appverifier pageheap instrumented crashes) is that the lambda is running twice -- both on the .NET GC thread, but also on the UI thread.

I don't see any way this could happen unless DispatcherQueue.TryEnqueue is returning false but then also running the lambda, but that also doesn't seem plausible from looking at the TryEnqueue code.

At this point we think we need to instrument this cleanup code to get closer to the source of the problem. Initially I tried to rewrite this code so that we never delete on the GC thread, but there are legitimate scenarios where a .NET caller could hold objects alive beyond the lifetime of the UI thread / DispatcherQueue, and it would be best if we don't leak in those scenarios.

Just to keep things simple, I've changed the lambda to keep track of the state when it was run so that if it in gets run a second time we'll have information about what happened the first time through. Also this will crash instead of corrupting the heap.

The simplest way to do this is to allocate another blob of shared state to track this -- I'm open to suggestions about how to do this without needing another heap allocation though.

### CODE REVIEW

**Comment by benwestprog** [comment]
>         bool queued = false;

Remove?

---
Refers to: dev/inc/RuntimeClassHelpers.h:137 in fe7d30a. [](commit_id = fe7d30a76aeb3819055194d3bcb41a47b85104a3, deletion_comment = False)

File: dev/inc/RuntimeClassHelpers.h [hunk 1]
```diff
@@ -139,19 +139,30 @@ struct ReferenceTracker : public ImplT<D, I ..., ::IReferenceTrackerExtension>,
         // See if we're on the UI thread
         if(!self->IsOnThread())
         {
+            struct LoggingState
+            {
+                void* dispatcherQueueWhenLambdaRan{ nullptr };
```
Comment (benwestprog) [inline dev/inc/RuntimeClassHelpers.h:144]: >dispatcherQueueWhenLambdaRan [](http://example.com/codeflow?start=22&length=28)

Unlikely to matter, but record original coreDispatcher too?

**Review by benwestprog — APPROVED** [review]
:shipit:

File: dev/inc/RuntimeClassHelpers.h [hunk 2]
```diff
@@ -134,24 +134,35 @@ struct ReferenceTracker : public ImplT<D, I ..., ::IReferenceTrackerExtension>,
...
+                void* coreDispatcherWhenLambdaRan{ nullptr };
+                std::atomic<int> runCount;
+            };
+            auto loggingState = std::make_shared<LoggingState>();
```
Comment (kmahone) [inline dev/inc/RuntimeClassHelpers.h:146]: I think this could be achieved with a mutable lambda without having to create a shared pointer.

**Comment by kmahone** [comment]
There were some test failures in the previous Pipeline run, but that run has since been cleaned up so it is not possible to see what specifically has failed. I've re-run the Pipeline so we can get new results.

**Comment by kmahone** [comment]
WebView2Tests.EdgeProcessFailedTest is failing. I'll take a look.

**Comment by kmahone** [comment]
Looks like this WebView2 test is failing in branches without this change. So this change did not break it.

**Review by kmahone — APPROVED** [review]

**Comment by pubtom** [comment]
> We're seeing sporadic double delete crashes in the vicinity of ReferenceTracker's DeleteInstanceOnUIThread cleanup lambda. My best guess (supported by some appverifier pageheap instrumented crashes) is that the lambda is running twice -- both on the .NET GC thread, but also on the UI thread.

Is it possible that my heap corruption https://github.com/microsoft/microsoft-ui-xaml/issues/7917 issue related to this issue?

**Comment by kmahone** [comment]
> > We're seeing sporadic double delete crashes in the vicinity of ReferenceTracker's DeleteInstanceOnUIThread cleanup lambda. My best guess (supported by some appverifier pageheap instrumented crashes) is that the lambda is running twice -- both on the .NET GC thread, but also on the UI thread.
> 
> Is it possible that my heap corruption #7917 issue related to this issue?

It is definitely possible. We still don't understand the source of these errors, but hopefully this change will allow us to get better insight.

**Pull request merged**